### PR TITLE
SwiftUI: base button and font styles

### PR DESCRIPTION
**Pull request by fespinoza**
# Why?

To keep providing a base to develop with SwiftUI for FINN.

# What?

- Add `.finnFont(labelStyle)` to set swiftUI fonts with the FINN theme
- Add some common button styles
  - Default
  - Call To Action
  - Flat
- add `roundedBorder` View extension

# Show me

It looks like this. I had some issues in order to attempt to add this preview to FinniversKit

Screenshot 2020-05-06 at 15 57 10

<details>
  <summary>Sample Usage</summary>
  

```swift
struct ButtonUsage: View {
    var body: some View {
        VStack(spacing: .spacingL) {
            Text("Buttons").font(.largeTitle)

            VStack(alignment: .leading, spacing: .spacingL) {
                Text("Call to Action").font(.headline)

                VStack(spacing: .spacingS) {
                    HStack(spacing: .spacingM) {
                        Button("Normal") {}
                        .buttonStyle(CallToAction())

                        Button("Disabled") {}
                        .buttonStyle(CallToAction(background: .btnDisabled))
                    }

                    HStack(spacing: .spacingM) {
                        Button("Normal") {}
                        .buttonStyle(CallToAction(size: .small))

                        Button("Disabled") {}
                        .buttonStyle(CallToAction(size: .small, background: .btnDisabled))
                    }
                }

                Text("Default").font(.headline)

                VStack(spacing: .spacingS) {
                    HStack(spacing: .spacingM) {
                        Button("Normal") {}
                        .buttonStyle(DefaultStyle())

                        Button("Disabled") {}
                        .disabled(true)
                        .buttonStyle(DefaultStyle(textColor: .btnDisabled))
                    }

                    HStack(spacing: .spacingM) {
                        Button("Normal Small") {}
                        .buttonStyle(DefaultStyle(size: .small))

                        Button("Disabled Small") {}
                        .disabled(true)
                        .buttonStyle(DefaultStyle(size: .small, textColor: .btnDisabled))
                    }
                }

                Text("Flat").font(.headline)

                VStack(spacing: .spacingS) {
                    HStack(spacing: .spacingM) {
                        Button("Normal") {}
                        .buttonStyle(FlatStyle())

                        Button("Disabled") {}
                        .disabled(true)
                        .buttonStyle(FlatStyle(textColor: .btnDisabled))
                    }

                    HStack(spacing: .spacingM) {
                        Button("Normal Small") {}
                        .buttonStyle(FlatStyle(size: .small))

                        Button("Disabled Small") {}
                        .disabled(true)
                        .buttonStyle(FlatStyle(size: .small, textColor: .btnDisabled))
                    }
                }

                Text("Inline Flat").font(.headline)

                VStack(alignment: .leading, spacing: .spacingS) {
                    HStack(spacing: .spacingM) {
                        Button("Normal") {}
                        .buttonStyle(InlineFlatStyle())

                        Button("Disabled") {}
                        .disabled(true)
                        .buttonStyle(InlineFlatStyle(textColor: .btnDisabled))
                    }

                    HStack(spacing: .spacingM) {
                        Button("Normal Small") {}
                        .buttonStyle(InlineFlatStyle(size: .small))

                        Button("Disabled Small") {}
                        .disabled(true)
                        .buttonStyle(InlineFlatStyle(size: .small, textColor: .btnDisabled))
                    }
                }

                Spacer()
            }
        }
        .padding()
    }
}
```
</details>

### CODE REVIEW

File: FinniversKit/Sources/Base Components/ButtonStyle.swift [hunk 1]
```diff
@@ -0,0 +1,105 @@
+//
+//  Copyright © 2020 FINN AS. All rights reserved.
+//
+
+import SwiftUI
+
+@available(iOS 13.0, *)
+public struct InlineFlatStyle: ButtonStyle {
+    let font: Font
```
Comment (vadymmarkov) [inline FinniversKit/Sources/Base Components/ButtonStyle.swift:9]: Minor thing, but I think properties here and in the other structs below could be private.

File: FinniversKit/Sources/Base Components/ButtonStyle.swift [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        HStack {
+            Spacer()
+            configuration.label
+                .font(font)
+                .foregroundColor(textColor)
+            Spacer()
+        }
+        .padding(.vertical, .spacingS)
+        .padding(.horizontal, .spacingM)
+        .background(Color.bgPrimary)
```
Comment (vadymmarkov) [inline FinniversKit/Sources/Base Components/ButtonStyle.swift:69]: It seems like everything up until setting the background is the same as in `FlatStyle`. What do you think about doing something like:
```swift
FlatStyle(size: size, textColor: textColor)
   .makeBody(configuration: configuration)
   .background(Color.bgPrimary)
   .roundedBorder(radius: .spacingS, color: .btnDisabled)
```

Or perhaps there's even better way of reusing this code.

File: FinniversKit/Sources/Base Components/ButtonStyle.swift [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        .roundedBorder(radius: .spacingS, color: .btnDisabled)
+    }
+}
+
+@available(iOS 13.0, *)
+struct CallToAction: ButtonStyle {
+    let background: UIColor
+    let font: Font
+
+    init(size: Button.Size = .normal, background: UIColor = .btnPrimary) {
```
Comment (vadymmarkov) [inline FinniversKit/Sources/Base Components/ButtonStyle.swift:79]: Why `UIColor`?

File: FinniversKit/Sources/DNA/Color/Color+SwiftUI.swift [hunk 4]
```diff
@@ -4,6 +4,8 @@
 
 import SwiftUI
 
+// swiftlint:disable opening_brace
```
Comment (vadymmarkov) [inline FinniversKit/Sources/DNA/Color/Color+SwiftUI.swift:7]: What is it complaining about?

File: FinniversKit/Sources/Base Components/ButtonStyle.swift (hunk 2 above)
Comment (fespinoza) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:69]: I wrote the code initially and now I was looking for ways of reusing. Nice suggestion!

File: FinniversKit/Sources/Base Components/ButtonStyle.swift (hunk 3 above)
Comment (fespinoza) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:79]: In order to change the background when the button is pressed I need to call `UIColor.withAlphaComponent`. I didn't see a way to do this with `Color`

File: FinniversKit/Sources/DNA/Color/Color+SwiftUI.swift (hunk 4 above)
Comment (fespinoza) [reply FinniversKit/Sources/DNA/Color/Color+SwiftUI.swift:7]: it's about the spacing between the variable definitions and the braces

```swift
var foo: Color                    { Color.red }
var longVarName: Color  { Color.blue }
```

It complains between the extra spacing in `foo`. Which I just used more as a visual style thing in the code

File: FinniversKit/Sources/Base Components/ButtonStyle.swift (hunk 3 above)
Comment (vadymmarkov) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:79]: How about `Color.opacity`?

Comment (fespinoza) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:79]: I was looking for alpha component :P

**Comment by fespinoza** [comment]
@vadymmarkov this is the error I see when trying to add the preview to FinniversKit

<details>
  <summary>Error Message</summary>
  

```
'ButtonStyleUsageDemoView_Previews' is not a member type of 'FinniversKit'

----------------------------------------

CompileDylibError: Failed to build ButtonStyle+Previews.swift

Compiling failed: 'ButtonStyleUsageDemoView_Previews' is not a member type of 'FinniversKit'

/Users/fespinoza/Library/Developer/Xcode/DerivedData/FinniversKit-cwqvnnqupgopzfbkiumgbidoffyw/Build/Intermediates.noindex/Previews/Demo/Intermediates.noindex/FinniversKit.build/Debug-iphonesimulator/FinniversKit.build/Objects-normal/x86_64/ButtonStyle+Previews.2.preview-thunk.swift:115:86: error: 'ButtonStyleUsageDemoView_Previews' is not a member type of 'FinniversKit'
@available(iOS 13.0.0, *) typealias ButtonStyleUsageDemoView_Previews = FinniversKit.ButtonStyleUsageDemoView_Previews
                                                                        ~~~~~~~~~~~~ ^
/Users/fespinoza/Code/work/finn/FinniversKit/FinniversKit/Sources/FinniversKit.swift:8:20: note: 'FinniversKit' declared here
@objc public class FinniversKit: NSObject {
                   ^
/Users/fespinoza/Library/Developer/Xcode/DerivedData/FinniversKit-cwqvnnqupgopzfbkiumgbidoffyw/Build/Intermediates.noindex/Previews/Demo/Intermediates.noindex/FinniversKit.build/Debug-iphonesimulator/FinniversKit.build/Objects-normal/x86_64/ButtonStyle+Previews.2.preview-thunk.swift:114:77: error: 'ButtonStyleUsageDemoView' is not a member type of 'FinniversKit'
@available(iOS 13.0.0, *) typealias ButtonStyleUsageDemoView = FinniversKit.ButtonStyleUsageDemoView
                                                               ~~~~~~~~~~~~ ^
/Users/fespinoza/Code/work/finn/FinniversKit/FinniversKit/Sources/FinniversKit.swift:8:20: note: 'FinniversKit' declared here
@objc public class FinniversKit: NSObject {
                   ^
```
</details>

**Comment by fespinoza** [comment]
it seems this error is related to name crashes ☝️ 

https://github.com/feedback-assistant/reports/issues/52

I guess it's something with `SwiftUI.Button` and `FinniversKit.Button`

**Comment by fespinoza** [comment]
@vadymmarkov can you look it again?

File: FinniversKit/Sources/Base Components/ButtonStyle.swift [hunk 5]
```diff
@@ -0,0 +1,109 @@
+//
+//  Copyright © 2020 FINN AS. All rights reserved.
+//
+
+import SwiftUI
+
+@available(iOS 13.0, *)
+public struct InlineFlatStyle: ButtonStyle {
```
Comment (vadymmarkov) [inline FinniversKit/Sources/Base Components/ButtonStyle.swift:8]: Should we perhaps move button styles to FinnUI? I know it's not the right place for it, but at least we can keep our SwiftUI views and other components in one place. In general, I feel like it's time to introduce the new framework.

File: FinniversKit/Sources/Base Components/ButtonStyle.swift [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    public init(size: Button.Size = .normal, textColor: Color = .btnPrimary) {
+        self.size = size
+        self.font = size == .normal ? .finnFont(.bodyStrong) : .finnFont(.detailStrong)
+        self.textColor = textColor
+    }
+
+    public func makeBody(configuration: Configuration) -> some View {
+        HStack {
+            Spacer()
+            configuration
```
Comment (vadymmarkov) [inline FinniversKit/Sources/Base Components/ButtonStyle.swift:64]: Can we also use `InlineFlatStyle` here?

File: FinniversKit/Sources/Base Components/ButtonStyle.swift [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                .label
+                .font(font)
+                .foregroundColor(textColor)
+            Spacer()
+        }
+        .padding(.vertical, .spacingS)
+        .padding(.horizontal, .spacingM)
+        .background(
+            configuration.isPressed ? Color(.defaultButtonHighlightedBodyColor) : Color.bgPrimary
```
Comment (vadymmarkov) [inline FinniversKit/Sources/Base Components/ButtonStyle.swift:73]: Have you tested this in dark mode? I got some weird blinking with using `defaultButtonHighlightedBodyColor` on press in basic list cells.

File: FinniversKit/Sources/Base Components/ButtonStyle.swift (hunk 6 above)
Comment (vadymmarkov) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:64]: Or perhaps even `FlatStyle` since it also has `HStack` with the same paddings.

Comment (fespinoza) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:64]: I had it like that before, but the highlighted state for FlatStyle is different than for the DefaultStyle

File: FinniversKit/Sources/Base Components/ButtonStyle.swift (hunk 7 above)
Comment (fespinoza) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:73]: yes, it works well

File: FinniversKit/Sources/Base Components/ButtonStyle.swift (hunk 6 above)
Comment (vadymmarkov) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:64]: Ok, then never mind 😉

File: FinniversKit/Sources/Base Components/ButtonStyle.swift (hunk 5 above)
Comment (fespinoza) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:8]: Mmm, I am not sure, we could.

If so, then we should not have swiftUI code within FinniversKit at all. Maybe just depend on it to get the assets/fonts, meaning that Color and Font overrides should be part of FinnUI too.

In that sense, I would keep the stuff in FinniversKit in this PR and move them in a new one, when we have a clear
plan for it

Comment (vadymmarkov) [reply FinniversKit/Sources/Base Components/ButtonStyle.swift:8]: What this new SwiftUI framework needs from FinniversKit is just he DNA part like fonts, colours and assets. It's doesn't really need UIKit components like Button and Label. That would probably resolve some of the conflicts like Finnivers/SwiftUI Button type. 

But yes, I agree. Let's discuss this further and prepare a clear plan for it.

**Pull request merged**